serde-reflection/src/de.rs

```diff
@@ -391,55 +396,147 @@ impl<'de, 'a> de::Deserializer<'de> for Deserializer<'de, 'a> {
-        // incomplete already, pick the first index.
-        let index = if known_variants.len() == variants.len()
-            || self.tracer.incomplete_enums.contains(name)
-        {
-            0
-        } else {
-            let mut index = known_variants.len() as u32;
-            // Scan the range 0..=known_variants.len() downwards to find the next
-            // variant index to explore.
-            while known_variants.contains_key(&index) {
-                index -= 1;
+
+        // If the enum is marked as incomplete, just visit the first index
+        // because we presume it avoids recursion.
+        if self.tracer.incomplete_enums.contains_key(enum_name) {
+            return visitor.visit_enum(EnumDeserializer::new(
+                self.tracer,
+                self.samples,
+                VariantId::Index(0),
+                &mut VariantFormat::unknown(),
+            ));
+        }
+
+        // First visit each of the variants by name according to `variants`.
+        // Later revisit them by u32 index until an index matching each of the
+        // named variants has been determined.
+        let provisional_min = u32::MAX - (variants.len() - 1) as u32;
+        for (i, &variant_name) in variants.iter().enumerate() {
+            if !self
+                .tracer
+                .discriminants
+                .contains_key(&(enum_type_id, VariantId::Name(variant_name)))
+            {
+                // Insert into known_variants with a provisional index.
+                let provisional_index = provisional_min + i as u32;
+                let variant = known_variants
+                    .entry(provisional_index)
+                    .or_insert_with(|| Named {
+                        name: variant_name.to_owned(),
+                        value: VariantFormat::unknown(),
+                    });
+                self.tracer
+                    .incomplete_enums
+                    .insert(enum_name.into(), EnumProgress::NamedVariantsRemaining);
+                // Compute the discriminant and format for this variant.
+                let mut value = variant.value.clone();
+                let enum_value = visitor.visit_enum(EnumDeserializer::new(
+                    self.tracer,
+                    self.samples,
+                    VariantId::Name(variant_name),
+                    &mut value,
+                ))?;
+                let discriminant = Discriminant::of(&enum_value);
+                self.tracer
+                    .discriminants
+                    .insert((enum_type_id, VariantId::Name(variant_name)), discriminant);
+                return Ok(enum_value);
+            }
+        }
+
+        // We know the discriminant for every variant name. Now visit them again
+        // by index to find the u32 id that goes with each name.
+        //
+        // If there are no provisional entries waiting for an index, just go
+        // with index 0.
```

Comment:
See comment above. It would be nice to not reach that line.

Reply:
Every call to `deserialize_enum` has to either call `visit_enum` or return an error. There is no existing short-circuit mechanism in serde-reflection (every deserialization error is propagated to the caller of `trace_type`). One can be added by introducing a kind of error that Tracer knows how to suppress, but that seems orthogonal to this PR.